### PR TITLE
Integrate RADE into FreeDV application

### PR DESCRIPTION
**Pull request by tmiw**
This PR is a reimplementation of #719 using the [C API](https://github.com/drowe67/radae/pull/27) and a much easier install path. 

TODO:

- [x] Install minimal Python when using the FreeDV installer
- [x] Create script to install pip and PyTorch, etc. 
- [x] Integrate script with FreeDV installer
- [x] Remove packages on uninstall
- [x] Update FreeDV code to support RADE using the C API
- [x] Add macOS support

### CODE REVIEW

File: cmake/rade-requirements.txt
Comment (tmiw) [inline cmake/rade-requirements.txt:1]: @drowe67, do we have any specific versions of the required prerequisites, or is the latest for all of them fine?

File: src/pipeline/RADEReceiveStep.cpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+{
+    return 8000; 
+}
+
+int RADEReceiveStep::getOutputSampleRate() const
+{
+    return 16000;
+}
+
+std::shared_ptr<short> RADEReceiveStep::execute(std::shared_ptr<short> inputSamples, int numInputSamples, int* numOutputSamples)
```
Comment (tmiw) [inline src/pipeline/RADEReceiveStep.cpp:66]: @drowe67, can you take a look at this function and let me know if it seems reasonable? I'm getting the following segfault on Linux after I push Start in the freedv-gui built by this branch:

```
encoder: 937200 weights
decoder: 907764 weights
encoder: 937200 weights
decoder: 907764 weights
Rs: 33.33 Rs': 50.00 Ts': 0.020 Nsmf: 120 Ns:   4 Nc:  30 M: 160 Ncp: 32
Input BPF bandwidth: 1740.000244 centre: 1475.000000
n_features_out: 432 nin_max: 1120 nin: 960
Setting callsign to K6AQ
[New Thread 0x7fff71a006c0 (LWP 4734)]
[Thread 0x7fff71a006c0 (LWP 4734) exited]
[New Thread 0x7fff71a006c0 (LWP 4735)]
[New Thread 0x7fff710006c0 (LWP 4736)]
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to record device alsa_input.pci-0000_02_02.0.analog-stereo
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to playback device alsa_output.platform-snd_aloop.0.analog-stereo
[New Thread 0x7fff63e006c0 (LWP 4737)]
[New Thread 0x7fff634006c0 (LWP 4738)]
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to record device alsa_input.platform-snd_aloop.0.analog-stereo
PulseAudioDevice[Sat 05 Oct 2024 11:15:24 PM PDT]: connecting to playback device alsa_output.pci-0000_02_02.0.analog-stereo
[New Thread 0x7fff62a006c0 (LWP 4739)]
[Thread 0x7fffc7e006c0 (LWP 4719) exited]
txRxThread: timeout while waiting for CV, tx = 1
txRxThread: timeout while waiting for CV, tx = 0
txRxThread: timeout while waiting for CV, tx = 1
txRxThread: timeout while waiting for CV, tx = 0
txRxThread: timeout while waiting for CV, tx = 1

Thread 31 "FreeDV rxThread" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff634006c0 (LWP 4738)]
0x00007ffff44f78d5 in ?? () from /lib/x86_64-linux-gnu/libpython3.12.so.1.0
(gdb) bt
#0  0x00007ffff44f78d5 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#1  0x00007ffff449c936 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#2  0x00007ffff449d356 in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#3  0x00007ffff438604d in ??? () at /lib/x86_64-linux-gnu/libpython3.12.so.1.0
#4  0x00007ffff7e5f457 in rade_rx ()
    at /home/mooneer/freedv-gui/build_linux/rade_build/src/librade.so.0.1
#5  0x000055555565feb7 in RADEReceiveStep::execute
    (this=0x7fff38010a60, inputSamples=std::shared_ptr<short> (use count 6, weak count 0) = {...}, numInputSamples=141, numOutputSamples=0x7fff633fe6ac)
    at /home/mooneer/freedv-gui/src/pipeline/RADEReceiveStep.cpp:95
#6  0x0000555555650629 in operator()
    (__closure=0x7fff380390f8, inSamples=std::shared_ptr<short> (use count 6, weak count 0) = {...}, numSamples=160)
    at /home/mooneer/freedv-gui/src/pipeline/ParallelStep.cpp:215
#7  0x0000555555651cc0 in std::__invoke_impl<std::pair<std::shared_ptr<short int>, int>, ParallelStep::enqueueTask_(ThreadInfo*, IPipelineStep*, std::shared_ptr<short int>, int)::<lambda(std::shared_ptr<short int>, int)>&, std::shared_ptr<short int>, int>(std::__invoke_other, struct {...} &) (__f=...)
    at /usr/include/c++/13/bits/invoke.h:61
#8  0x0000555555651a7c in std::__invoke_r<std::pair<std::shared_ptr<short int>, int>, ParallelStep::enqueueTask_(ThreadInfo*, IPipelineStep*, std::shared_ptr<short int>, int)::<lambda(std::shared_ptr<short int>, int)>&, std::shared_ptr<short int>, int>(struct {...} &) (__fn=...)
--Type <RET> for more, q to quit, c to continue without paging--q
```

I'm executing by running the following:

```
mooneer@ubuntu-vm:~/freedv-gui/build_linux$ . ~/radae/radae-venv/bin/activate
(radae-venv) mooneer@ubuntu-vm:~/freedv-gui/build_linux$ PYTHONPATH=rade_src/embed:rade_src/ src/freedv
```

Comment (tmiw) [reply src/pipeline/RADEReceiveStep.cpp:66]: Actually, looks like it was a threading issue. See https://github.com/drowe67/radae/pull/27/commits/300f2e302298577efac78748c941a3d5433a931b.

Unfortunately RADE seems to be behaving pretty slowly in the Ubuntu VM I'm using for some reason. I need to investigate this more.

**Comment by barjac** [comment]
This branch package build is failing trying to access gitlab for opus and github for radae, however it is also failing to detect that opus is already installed from our package in the build chroot.

I expect any downloads to fail during a package build on our official build servers as this is not allowed but not on mine.

This brings me to a question.
I have been looking into packaging radae but it seems that it's dependencies are very processor dependant and one size does not fit all.

If builds must be optimized for each machine, then packaging becomes impractical.

Or is this not so critical?

**Comment by Tyrbiter** [comment]
Bit of a work in progress I think, I have just used the build_linux.sh approach and tried to run the modem in RADE mode, I got this:

model file: TBD
import_array returned: 0
ModuleNotFoundError: No module named 'radae_txe'
Error: importing radae_txe

Probably something python related that I have not installed and to be honest it's too soon to be worried about it now as @tmiw is busy getting a working test binary for users.

**Comment by barjac** [comment]
> Bit of a work in progress I think, I have just used the build_linux.sh approach 
That borks at finding python3 just now.

> 
> Probably something python related that I have not installed and to be honest it's too soon to be worried about it now as @tmiw is busy getting a working test binary for users.

Yes I realize it's too soon - I was simply trying to prepare the groundwork for packaging.

**Comment by tmiw** [comment]
IIRC this should get you a working version on Linux with this PR if you'd like to try it out:

```
$ python3 -m venv rade-venv
$ . ./rade-venv/bin/activate
(rade-venv) $ pip install -r cmake/rade-requirements.txt
(rade-venv) $ ./build_linux.sh
...
(rade-venv) $ cd build_linux
(rade-venv) $ ln -s `pwd`/rade_src/model19_check3 model19_check3
(rade-venv) $ PYTHONPATH=`pwd`/rade_src src/freedv
```

Any suggestions that would make future packaging easier are definitely appreciated. (Currently working through trying to get a macOS version built and packaged and that's looking to be a giant pain so far.)

**Comment by Tyrbiter** [comment]
Falls over for me at the pip install stage:

(rade-venv) bdm@deangelis:~/git/freedv-gui-ms-rade-integ$ pip install -r cmake/rade-requirements.txt
ERROR: Could not find a version that satisfies the requirement torch (from versions: none)
ERROR: No matching distribution found for torch

I have these rpms installed for python pip

python-pip-wheel-24.2-1.fc41.noarch
python3-pip-24.2-1.fc41.noarch
python3-pip-requirements-parser-32.0.1-5.fc41.noarch
python3-pipdeptree-2.23.1-1.fc41.noarch

and these for torch and pytorch

torch-3.1-5.1.x86_64
python3-torch-2.4.0-7.fc41.x86_64

It's not a big hassle, I was only seeing if I could try it out but this might help with overall packaging requirements.

**Comment by tmiw** [comment]
> Falls over for me at the pip install stage:
> 
> (rade-venv) bdm@deangelis:~/git/freedv-gui-ms-rade-integ$ pip install -r cmake/rade-requirements.txt ERROR: Could not find a version that satisfies the requirement torch (from versions: none) ERROR: No matching distribution found for torch

Maybe append `--index-url https://download.pytorch.org/whl/cpu` to that command line? At least according to https://pytorch.org/get-started/locally/ anyway.

> I have these rpms installed for python pip
> 
> python-pip-wheel-24.2-1.fc41.noarch python3-pip-24.2-1.fc41.noarch python3-pip-requirements-parser-32.0.1-5.fc41.noarch python3-pipdeptree-2.23.1-1.fc41.noarch
> 
> and these for torch and pytorch
> 
> torch-3.1-5.1.x86_64 python3-torch-2.4.0-7.fc41.x86_64

I think if the PyTorch packages are already installed, pip isn't needed if you can also find a matplotlib Fedora package. If not, you'll need to install the latter using `pip` before proceeding.

**Comment by Tyrbiter** [comment]
This is what I have for any package with matplotlib in the name:

python3-matplotlib-data-fonts-3.9.1-2.fc41.noarch
python3-matplotlib-data-3.9.1-2.fc41.noarch
python3-matplotlib-inline-0.1.7-3.fc41.noarch
python3-matplotlib-tk-3.9.1-2.fc41.x86_64
python3-matplotlib-3.9.1-2.fc41.x86_64

have just installed all of the python3-matplotlib packages, for -wx, -gtk* and -qt*

I will try the other suggestion for the pip install failure.

**Comment by Tyrbiter** [comment]
Now ./build_linux.sh reports this error:

-- Will build opus with FARGAN
CMake Error at /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find Python3 (missing: Python3_NumPy_INCLUDE_DIRS NumPy) (found
  version "3.13.0")
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake/Modules/FindPython/Support.cmake:3867 (find_package_handle_standard_args)
  /usr/share/cmake/Modules/FindPython3.cmake:545 (include)
  CMakeLists.txt:50 (find_package)


-- Configuring incomplete, errors occurred!
make[2]: *** [CMakeFiles/build_rade.dir/build.make:92: build_rade-prefix/src/build_rade-stamp/build_rade-configure] Error 1
make[2]: Leaving directory '/home/bdm/git/freedv-gui-ms-rade-integ/build_linux'

I'm trying to work out how to get the Python3_NumPy_INCLUDE_DIRS to be found. So far I can't find an appropriate -devel package.

**Comment by tmiw** [comment]
> I'm trying to work out how to get the Python3_NumPy_INCLUDE_DIRS to be found. So far I can't find an appropriate -devel package.

If you run `pip install numpy` while in that venv, does that help?

**Comment by Tyrbiter** [comment]
> > I'm trying to work out how to get the Python3_NumPy_INCLUDE_DIRS to be found. So far I can't find an appropriate -devel package.
> 
> If you run `pip install numpy` while in that venv, does that help?

I don't think it will, the Fedora numpy source rpm package builds and installs these:

python3-numpy-1.26.4-8.fc41.x86_64
python3-numpy-f2py-1.26.4-8.fc41.x86_64

there are no -devel packages for numpy. The include files are here:

/usr/lib64/python3.13/site-packages/numpy/core/include/numpy/

and the files in that directory are these:

$ ll /usr/lib64/python3.13/site-packages/numpy/core/include/numpy/
total 380
-rw-r--r--. 1 root root   282 Jul 18 01:00 arrayobject.h
-rw-r--r--. 1 root root  3944 Jul 18 01:00 arrayscalars.h
-rw-r--r--. 1 root root 16697 Jul 18 01:00 _dtype_api.h
-rw-r--r--. 1 root root 15532 Jul 18 01:00 experimental_dtype_api.h
-rw-r--r--. 1 root root  1959 Jul 18 01:00 halffloat.h
-rw-r--r--. 1 root root 12116 Jul 18 01:00 __multiarray_api.c
-rw-r--r--. 1 root root 61450 Jul 18 01:00 __multiarray_api.h
-rw-r--r--. 1 root root 10183 Jul 18 01:00 ndarrayobject.h
-rw-r--r--. 1 root root 68009 Jul 18 01:00 ndarraytypes.h
-rw-r--r--. 1 root root  1857 Jul 18 01:00 _neighborhood_iterator_imp.h
-rw-r--r--. 1 root root  6830 Jul 18 01:00 noprefix.h
-rw-r--r--. 1 root root  4327 Jul 18 01:00 npy_1_7_deprecated_api.h
-rw-r--r--. 1 root root 15883 Jul 18 01:00 npy_3kcompat.h
-rw-r--r--. 1 root root 37746 Jul 18 01:00 npy_common.h
-rw-r--r--. 1 root root  4629 Jul 18 01:00 npy_cpu.h
-rw-r--r--. 1 root root  2786 Jul 18 01:00 npy_endian.h
-rw-r--r--. 1 root root  1948 Jul 18 01:00 npy_interrupt.h
-rw-r--r--. 1 root root 18945 Jul 18 01:00 npy_math.h
-rw-r--r--. 1 root root   678 Jul 18 01:00 npy_no_deprecated_api.h
-rw-r--r--. 1 root root  1256 Jul 18 01:00 npy_os.h
-rw-r--r--. 1 root root   855 Jul 18 01:00 _numpyconfig.h
-rw-r--r--. 1 root root  5308 Jul 18 01:00 numpyconfig.h
-rw-r--r--. 1 root root  6405 Jul 18 01:00 old_defines.h
drwxr-xr-x. 1 root root    68 Sep 14 17:04 random
-rw-r--r--. 1 root root  1714 Jul 18 01:00 __ufunc_api.c
-rw-r--r--. 1 root root 12456 Jul 18 01:00 __ufunc_api.h
-rw-r--r--. 1 root root 12070 Jul 18 01:00 ufuncobject.h
-rw-r--r--. 1 root root  1185 Jul 18 01:00 utils.h

I think it ought to be found but something is not adding that directory to the include search path.

**Comment by tmiw** [comment]
Maybe it's something with the version of `cmake`? I'm currently using 3.28.3.

**Comment by Tyrbiter** [comment]
> Maybe it's something with the version of `cmake`? I'm currently using 3.28.3 with the Fedora -7 package versions.

Same version installed here, however from the cmake source rpm build changelog:

* Wed May 15 2024 Sandro <devel@penguinpee.nl> - 3.28.3-5
- Add python3-cmake sub package providing metadata for Python packages
  depending on CMake to facilitate automatic dependency resolution.

I have now installed this package, which was not installed automatically. I will report what happens.

**Comment by tmiw** [comment]
I just tried on Fedora 40 (after running `dnf update` first) and it looks like it was able to detect Python correctly without needing venv. I installed `python3-matplotlib` and `python3-torch` beforehand FWIW.

Building RADE manually results in the following:

image

I noticed that I'm using Python 3.12 while you're using 3.13; maybe that's why there's issues?

BTW `numpy` definitely needs to be 2.x or bad things happen.

**Comment by Tyrbiter** [comment]
OK, well that is useful information. Python 3.13 comes with Fedora 41, which will release in a couple of weeks roughly.

Note that you will need the python3-setuptools package installed too in case things are not working.

I don't know the effect of venv, maybe that is like a chroot and prevents some parts of the standard install from being visible.

I will have a play and see what happens. Fedora has not yet packaged numpy 2.x as it seems there are a lot of dependencies that get broken and this close to a code freeze (Tuesday) there is no chance of any movement.

**Comment by Tyrbiter** [comment]
I now have the modem running in receive for RADE mode, it looks like the Fedora python3-torch package is OK but still needed to install the newer numpy using pip, I think that is persistent?

**Comment by tmiw** [comment]
> I now have the modem running in receive for RADE mode, it looks like the Fedora python3-torch package is OK but still needed to install the newer numpy using pip, I think that is persistent?

Yeah, pip installs in your home folder by default (unless running as root).

> I don't know where to make this change, but I assume that it would make things run faster in the python modules.

IIRC the printing of this warning is a bug. AFAIK it's still using SSE/AVX/etc.

**Comment by Tyrbiter** [comment]
> > I now have the modem running in receive for RADE mode, it looks like the Fedora python3-torch package is OK but still needed to install the newer numpy using pip, I think that is persistent?
> 
> Yeah, pip installs in your home folder by default (unless running as root).

Indeed, I found it in the rade-venv folder created the other day.

> 
> > I don't know where to make this change, but I assume that it would make things run faster in the python modules.
> 
> IIRC the printing of this warning is a bug. AFAIK it's still using SSE/AVX/etc.

I'm not overly worried, it's very early days to be worried about this sort of thing. I'm more concerned about ensuring it looks for the correct numpy version so will do some more investigation.

**Comment by barjac** [comment]
I have been battling with this monster for several weeks now and am grinding to a halt. :(
It seems that build_linux.sh in 2.0.0 has not been updated.
radae from github does not install anything, so packaging has stalled on that.

@tyrbiter Have you got anywhere? I even made an install of Fedora 40 to attempt to see if a build there would at least give me something to compare, since several new dependencies still do not build for me in Mageia. IIRC python-torch pulled in 600 packages in Fedora!
The only report that I have heard from a Windows user was a crash on starting the modem in radae mode.

**Comment by Tyrbiter** [comment]
The only thing I can say is that for now I am using the python-torch approach, but I do have Fedora 41 now, it's a couple of weeks away from release. The difficulty is that the python-numpy packages are at v 1.x so it's necessary to install numpy 2.x using pip, not sure how long until numpy 2.x is available for Fedora but I suspect it could take until Fedora 42 because of the huge number of dependent packages that will need updating.

I have python3-torch and python-torchaudio installed from the distro repos. build_linux.sh downloads the rade code and puts it in build_linux/rade_src and the numpy 2.x is in build_linux/rade-venv/lib/python3.13/site-packages/

As for packaging rade (as it's been renamed for ease of pronunciation) that is going to take some time because a port to C is a lot of work and has to be a rigorous process with unit tests and careful code optimisation.

I have a running version here, but it's not as simple as clicking the icon for freedv v1.x, needs a soft link and starting it from the right directory and telling it the correct python path.

**Comment by tmiw** [comment]
IIRC I think the macOS version actually uses numpy 1.x.  Not sure why that needed 1.x while the other platforms worked with 2.x /shrug

But yeah, probably premature to try to package this. If we can get some consistently working Linux instructions written, Linux users can at least compile it to try it out for the time being.

**Comment by Tyrbiter** [comment]
> IIRC I think the macOS version actually uses numpy 1.x. Not sure why that needed 1.x while the other platforms worked with 2.x /shrug

Is there not something in 2.x that is needed by rade? Seems odd that it works on MacOS.

**Comment by tmiw** [comment]
> > IIRC I think the macOS version actually uses numpy 1.x. Not sure why that needed 1.x while the other platforms worked with 2.x /shrug
> 
> Is there not something in 2.x that is needed by rade? Seems odd that it works on MacOS.

It might be more that it has to be compiled with the same version as what it's going to use at run time, and that may be easier with 1.x especially with how universal binaries work. Not sure to be honest.

**Comment by Tyrbiter** [comment]
No point worrying, eventually numpy 2.x will get to MacOS.

**Comment by tmiw** [comment]
> No point worrying, eventually numpy 2.x will get to MacOS.

I think there is a 2.x on macOS. I'm also using a slightly older version of torch (due to not wanting to only support ARM) so that may be contributing to the issue too.

**Comment by tmiw** [comment]
Going to merge this PR into v2.0-dev since this is getting pretty big (and the basic functionality works). Additional PRs will be created on v2.0-dev to resolve bugs/make it easier for the PLT to review.

**Pull request merged**